gateway/gateway.go

```diff
@@ -17,34 +17,122 @@ type Config struct {
+type dummyGetOpts struct{}
 
-	// GetBlock return a block from a certain CID.
-	GetBlock(context.Context, cid.Cid) (blocks.Block, error)
+var CommonGetOptions dummyGetOpts
+
+// GetRange is a range request for some file data
+// Note: currently a full file object should be returned but reading from values outside the range can error
+func (o dummyGetOpts) GetRange(from, to int) GetOpt {
+	return func(options *GetOptions) error {
+		options.RangeFrom = from
+		options.RangeTo = to
+		return nil
+	}
+}
+
+// GetFullDepth fetches all data linked from the last logical node in the path. In particular, recursively fetch an
+// entire UnixFS directory.
+func (o dummyGetOpts) GetFullDepth() GetOpt {
+	return func(options *GetOptions) error {
+		options.FullDepth = true
+		return nil
+	}
+}
+
+// GetRawBlock fetches the data at the last path element as if it were a raw block rather than something more complex
+// such as a UnixFS file or directory.
+func (o dummyGetOpts) GetRawBlock() GetOpt {
+	return func(options *GetOptions) error {
+		options.RawBlock = true
+		return nil
+	}
+}
+
+// API TODO: We might need to define some sentinel errors here to help the Gateway figure out what HTTP Status Code to return
+// For example, distinguishing between timeout/failures to fetch vs receiving invalid data, vs impossible paths vs wrong data types, etc.
+type API interface {
+	// Get returns a file or directory depending on what the path is that has been requested.
+	// There are multiple options passable to this function, read them for more information.
+	Get(context.Context, ImmutablePath, ...GetOpt) (GatewayMetadata, files.Node, error)
+
+	// Head returns a file or directory depending on what the path is that has been requested.
+	// For UnixFS files should return a file where at least the first 1024 bytes can be read and has the correct file size
+	// For all other data types returning just size information is sufficient
+	// TODO: give function more explicit return types
+	Head(context.Context, ImmutablePath) (GatewayMetadata, files.Node, error)
+
+	// GetCAR returns a CAR file for the given immutable path
+	// Returns an initial error if there was an issue before the CAR streaming begins as well as a channel with a single
+	// that may contain a single error for if any errors occur during the streaming. If there was an initial error the
+	// error channel is nil
+	// TODO: Make this function signature better
+	GetCAR(context.Context, ImmutablePath) (GatewayMetadata, io.ReadCloser, error, <-chan error)
+
+	// IsCached returns whether or not the path exists locally.
+	IsCached(context.Context, path.Path) bool
 
 	// GetIPNSRecord retrieves the best IPNS record for a given CID (libp2p-key)
 	// from the routing system.
 	GetIPNSRecord(context.Context, cid.Cid) ([]byte, error)
 
+	// ResolveMutable takes a mutable path and resolves it into an immutable one. This means recursively resolving any
+	// DNSLink or IPNS records.
```

Comment:
Reduce ambiguity what we mean by recursive (we dont want /ipns/id/foo/bar → /ipfs/bar-cid, right?)
```suggestion
	// ResolveMutable takes a mutable path and resolves it into an immutable one. This means recursively resolving any
	// /ipns/id (DNSLink or IPNS record identifier), but keeping all the remainder path:
	// /ipns/id/foo/bar → /ipfs/cid/foo/bar
```

Reply:
Used different text, we can see if that needs improving as well